### PR TITLE
feat(raster-loader) Update raster-loader to generate new Raster and Metadata table format

### PR DESCRIPTION
**Pull request by luipir**
Story details: https://app.shortcut.com/cartoteam/story/358996

### CODE REVIEW

File: raster_loader/io.py [hunk 1]
```diff
@@ -446,27 +303,12 @@ def raster_band_type(raster_dataset: rasterio.io.DatasetReader, band: int) -> st
...
-            ("lat_NE", "FLOAT", "NULLABLE"),
-            ("lon_SE", "FLOAT", "NULLABLE"),
-            ("lat_SE", "FLOAT", "NULLABLE"),
-            ("lon_SW", "FLOAT", "NULLABLE"),
-            ("lat_SW", "FLOAT", "NULLABLE"),
-            ("geog", "GEOGRAPHY", "NULLABLE"),
-        ]
+def table_columns(bands: List[str]) -> List[Tuple[str, str, str]]:
+    columns = [
+        ("block", "INTEGER", "REQUIRED"),
```
Comment (Jesus89) [inline raster_loader/io.py:308]: "metadata" can be added in this list

File: raster_loader/io.py [hunk 2]
```diff
@@ -512,30 +355,27 @@ def rasterio_windows_to_records(
...
+
+    """Check if raster is quadbin compatible."""
+    if "GoogleMapsCompatible" != raster_info.get("Tags", {}).get(
+        "Tiling Scheme", {}
+    ).get("NAME"):
+        msg = (
+            "The input raster must be a GoogleMapsCompatible raster.\n"
+            "You can make your raster compatible "
+            "by converting it using the following command:\n"
+            "gdalwarp your_raster.tif -of COG "
```
Comment (Jesus89) [inline raster_loader/io.py:371]: Note to update the suggested command to

```
gdalwarp -of COG -co TILING_SCHEME=GoogleMapsCompatible -co COMPRESS=DEFLATE -co OVERVIEWS=NONE -co ADD_ALPHA=NO -co RESAMPLING=NEAREST <input_raster>.tif <output_raster>.tif
```

File: raster_loader/io.py [hunk 3]
```diff
@@ -561,85 +401,52 @@ def rasterio_windows_to_records(
...
-            if metadata["max_pixel_block_width_in_pixel"] < window.width:
-                metadata["max_pixel_block_width_in_pixel"] = window.width
-            if (
-                window.height != metadata["max_pixel_block_height_in_pixel"]
-                or window.width != metadata["max_pixel_block_width_in_pixel"]
-            ):
-                metadata["irregular_pixel_block_shape"] = True
+            # metadata["nb_pixel_blocks"] += 1
+            # metadata["nb_pixel"] += window.width * window.height
+            metadata["num_blocks"] += 1
```
Comment (Jesus89) [inline raster_loader/io.py:448]: This is not efficient, let's use the metadata to compute the num_blocks, so the num_pixels before hand

File: raster_loader/io.py [hunk 4]
```diff
@@ -1107,19 +909,20 @@ def table_creator(columns, clustering):
...
             metadata,
             append_records,
-            output_quadbin,
             project_id,
             dataset_id,
             table_id,
             client=client,
         )
 
+        # !!! outdated due the fact that raster windows is constant for each
```
Comment (Jesus89) [inline raster_loader/io.py:918]: Yes, it can be removed then

File: raster_loader/io.py [hunk 5]
```diff
@@ -1154,125 +957,23 @@ def table_creator(columns, clustering):
...
-                            ARRAY_AGG(DISTINCT TO_JSON_STRING(gdal_transform))
-                        ) = 1,
-                        ANY_VALUE(gdal_transform),
-                        NULL
-                    ) AS gdal_transform
-                    {fetch_resolution}
-                )))
-                FROM united
-            ) WHERE {location} IS NULL
+            ) WHERE {location} = 0
```
Comment (Jesus89) [inline raster_loader/io.py:976]: WHERE block = 0 can be used since there is no other location

File: raster_loader/io.py [hunk 6]
```diff
@@ -1352,45 +1058,31 @@ def get_block_dims(file_path: str) -> tuple:
...
-                CAST(JSON_VALUE(attrs, '$.block_area') AS FLOAT64)
-                / (block_height*block_width)
-              )
-            {from_blocks_source}
-        """
-        area_query = f"""
-            SELECT SUM(CAST(JSON_VALUE(attrs, '$.block_area') AS FLOAT64))
-            {from_blocks_source}
-        """
+def inject_areas_query(raster_table: str) -> str:
```
Comment (Jesus89) [inline raster_loader/io.py:1061]: I think this can be completely removed. The new metadata does not include the area

File: raster_loader/tests/test_io.py [hunk 7]
```diff
@@ -155,183 +61,24 @@ def test_array_to_quadbin_record():
...
-        "block_width",
-        "attrs",
-        "band_1_uint8",
-    ]
-
-    assert sorted(list(result.columns)) == sorted(expected_columns)
-
-    # TODO: select metadata row and check metadata contents
-    # TODO: select some block row and check contents
+    assert record["metadata"] is None
```
Comment (Jesus89) [inline raster_loader/tests/test_io.py:68]: Check also the new metadata not None

File: setup.cfg [hunk 8]
```diff
@@ -33,10 +33,11 @@ install_requires =
     db-dtypes>=1.0.5
     google-cloud-bigquery>=3.4.0
     pandas>=1.3.4
-    pyarrow>=10.0.1
+    ; pyarrow>=10.0.1
```
Comment (Jesus89) [inline setup.cfg:36]: Why this ; ?

**Review by Jesus89 — CHANGES_REQUESTED** [review]
Overall looks good. But there are some details to complete before merging:

Regarding the metadata: Why are `width` and `height` fields included

```json
{
  "resolution": 8,
  "bands": [
    "band_1_uint8"
  ],
  "bounds": [
    -4.218749999999985,
    48.92249926375824,
    -1.4062499999999951,
    50.73645513701064
  ],
  "center": [
    -2.812499999999989,
    49.82947720038444,
    8
  ],
  "width": 512,
  "height": 512,
  "block_width": 256,
  "block_height": 256,
  "num_blocks": 4,
  "num_pixels": 262144
}
```

Additionally, let's remove some leftover code and update the tests

File: raster_loader/io.py (hunk 3 above)
Comment (luipir) [reply raster_loader/io.py:448]: I just followed the same logic as before, not clear why was in this way if there are some corner case to manage. BTW I'll remove.

File: setup.cfg (hunk 8 above)
Comment (luipir) [reply setup.cfg:36]: ??? no idea why there is this change·... I'll check. tnx to review.

**Comment by luipir** [comment]
> Overall looks good. But there are some details to complete before merging:
> 
> Regarding the metadata: Why are `width` and `height` fields included

give the overall size of the original raster. To have this info, should be calculated with block and num of blocks.
IMHO is exactly as center that can be derived from centroid of BBOX.
Should I remove then `width` and `height`?

File: raster_loader/io.py (hunk 3 above)
Comment (Jesus89) [reply raster_loader/io.py:448]: I guess there must be a way to get the num blocks from the raster info

**Comment by Jesus89** [comment]
This is not used actually but as num_blocks or num_pixels, we can keep it as informative data :+1:

File: raster_loader/cli/bigquery.py [hunk 9]
```diff
@@ -24,9 +24,23 @@ def bigquery(args=None):
...
 @click.option(
-    "--chunk_size", help="The number of blocks to upload in each chunk.", default=100
+    "--band",
+    help="Band(s) within raster to upload. "
+    "Could repeat --band to specify multiple bands.",
+    default=[1],
+    multiple=True,
+)
+@click.option(
+    "--band_column_name",
```
Comment (Jesus89) [inline raster_loader/cli/bigquery.py:35]: I would rename it to `band_name`

File: raster_loader/io.py (hunk 3 above)
Comment (Jesus89) [reply raster_loader/io.py]: I've changed the formula using width and height

**Review by Jesus89 — COMMENTED** [review]
I've reviewed and tested the code. Found and fixed some issues:
- Rename `band_column_name` to `band_name`
- Fix progressbar count (using the latest job)
- Remove `input_crs`
- Remove `pseudo_planar`
- Remove dead code
- Recompute `num_blocks` and `num_pixels`

**Review by Jesus89 — CHANGES_REQUESTED** [review]
I've found also an important bug:
- The metadata is not incrementally updated on mosaic rasters. This must be done by reading the existing metadata and merging the old and new values.

There are some values missing in the metadata:
- minresolution
- maxresolution
- nodata

We should keep `wide_number_mode=>'round'` in PARSE_JSON. This was added to solve a recent bug

@luipir @volaya

**Comment by luipir** [comment]
> I've found also an important bug:
> 
> * The metadata is not incrementally updated on mosaic rasters. This must be done by reading the existing metadata and merging the old and new values.
> 
> There are some values missing in the metadata:
> 
> * minresolution
> * maxresolution
> * nodata
> 
> We should keep `wide_number_mode=>'round'` in PARSE_JSON. This was added to solve a recent bug
> 
> @luipir @volaya

I didn't understand well, could you point the mosaic raster you used, probably I missed a use case case that nor was covered by tests.

**Comment by luipir** [comment]
> * Fix progressbar count (using the latest job)

hmmm... not sure it is a reliable way to progress bar, because jobs are executed without clear execution order

**Comment by Jesus89** [comment]
> I didn't understand well @Jesus89 , could you point the mosaic raster you used, probably I missed a use case.

It's basically this code: https://github.com/CartoDB/raster-loader/pull/116/files#diff-35ccab082c7efa2b1a83c6fef7690c297582c4d96b4551043db8f491083886a2L1177. If we add two pieces of a raster mosaic, the first upload, the metadata will be associated to the first piece, but the second upload, the metadata will need to be updated to include both pieces

> hmmm... not sure it is a reliable way to progress bar, because jobs are executed without clear execution order

The bug was that it always uses the num_records of the latest job, instead of the record of the specific job. For example, for 1510 blocks, it will push 500 + 500 + 500 + 10, but the previous implementation added 500 + 10 + 10 + 10. It does not matter when the jobs end, but they need to add to the progress bar the correct amount.

> could these values be different by that get from rasterio metadata? if a block fail to upload fail all upload so IMHO no need recompute these data. Other if there is a use case I do not considered.

The num params are computed from the raster medatata width, height, block_width, block_height. It pushed the metadata for success upload, if the upload fails, it will have wrong metadata anyway

**Comment by volaya** [comment]
@Jesus89 I have fixed the issues that you mentioned. While adding a new test for the bounds re-calculation, I found another issue that I have fixed, related to metadata. It was not being written in some cases. Tests were passing because the  fixtures were actually wrong. 

Let me know what you think

**Comment by luipir** [comment]
LGTM btw Github do not allow me to approve a PR opened by me :)

File: raster_loader/io.py [hunk 10]
```diff
@@ -825,25 +846,34 @@ def write_metadata(
...
     if append_records:
+        """Requires bigquery."""
+        if not _has_bigquery:  # pragma: no cover
+            import_error_bigquery()
+
+        if client is None:
+            client = bigquery.Client(project=project_id)
+
+        old_metadata = get_metadata(project_id, dataset_id, table_id, client=client)
+        metadata["bounds"] = (
```
Comment (luipir) [inline raster_loader/io.py:857]: i it necessarry to update center too?

Comment (Jesus89) [reply raster_loader/io.py:857]: Yes

**Review by Jesus89 — APPROVED** [review]
Changes look good @volaya!

It seems you fixed all the reported bugs, added extra checks for mosaic rasters, and added more tests. The refactor of the code looks nice too, it's more clear now.

**Comment by Jesus89** [comment]
Let's rename "band_name" to "name" in metadata.bands @volaya. We will apply the changes in the API and AT @vdelacruzb

**Review by Jesus89 — APPROVED** [review]
